lib/uri_checker.rb

```diff
@@ -0,0 +1,213 @@
+require "uri"
+
+module UriChecker
+  class Report
+    attr_reader :errors, :warnings
+
+    def initialize(errors: nil, warnings: nil)
+      @errors = errors || Hash.new { |hash, key| hash[key] = [] }
+      @warnings = warnings || Hash.new { |hash, key| hash[key] = [] }
+    end
+
+    def merge!(other)
```

Comment:
they're unfortunately quite inconsistent, so while you're understanding is correct in some circumstances it's not always the case. The most consistent definition is that they offer a dangerous alternative of a method. 

So `downcase` (a lovely translation oddity) is one where you've got `downcase` non modifying and `downcase!` modifying

whereas if you take ActiveRecord `save` and `save!` the bang indicates that it'll throw an exception. 

Since the language is object oriented there are lots of places where state is changed by a method call too so can quickly become a bit of a nightmare applying the bang modify pattern, whereas it gets a lots simpler applied to just dangerous alternative